README.md
# clue/viewvc-api-react [![Build Status](https://travis-ci.org/clue/php-viewvc-api-react.svg?branch=master)](https://travis-ci.org/clue/php-viewvc-api-react)

Simple, async API-like access to your [ViewVC](http://viewvc.org/) web interface (Subversion/CVS browser), built on top of [React PHP](http://reactphp.org/).

**Table of Contents**

* [Quickstart example](#quickstart-example)
* [Usage](#usage)
  * [Client](#client)
    * [Actions](#actions)
    * [Promises](#promises)
    * [Blocking](#blocking)
    * [Streaming](#streaming)
* [Install](#install)
* [License](#license)

## Quickstart example

Once [installed](#install), you can use the following code to fetch a directory
listing from the given ViewVC URL:

```php
$loop = React\EventLoop\Factory::create();
$browser = new Clue\React\Buzz\Browser($loop);
$client = new Client($browser->withBase('http://example.com/viewvc/'));

$client->fetchDirectory('/')->then(function ($files) {
    echo 'Files: ' . implode(', ', $files) . PHP_EOL;
});

$loop->run();
```

See also the [examples](examples).

## Usage

### Client

The `Client` is responsible for assembling and sending HTTP requests to the remote ViewVC web interface.
It requires a [`Browser`](https://github.com/clue/php-buzz-react#browser) object
bound to the main [`EventLoop`](https://github.com/reactphp/event-loop#usage)
in order to handle async requests:

```php
$loop = React\EventLoop\Factory::create();
$browser = new Clue\React\Buzz\Browser($loop);

$client = new Client($browser->withBase('http://example.com/viewvc/'));
```

The `Client` API uses relative URIs to reference files and directories in your
ViewVC installation, so make sure to apply the base URI as depicted above.

If you need custom DNS or proxy settings, you can explicitly pass a
custom [`Browser`](https://github.com/clue/php-buzz-react#browser) instance.

#### Actions

ViewVC does not officially expose an API. However, its REST-like URLs make it
easy to construct the right requests and scrape the results from its HTML
output.
All public methods resemble these respective actions otherwise available in the
ViewVC web interface.

```php
$client->fetchDirectory($path, $revision = null);
$client->fetchFile($path, $revision = null);
$client->fetchPatch($path, $r1, $r2);
$client->fetchLog($path, $revision = null);

// many more…
```

All actions support async operation by returning [promises](#promises).

Listing all available actions is out of scope here, please refer to the [class outline](src/Client.php).

#### Promises

Sending requests is async (non-blocking), so you can actually send multiple requests in parallel.
ViewVC will respond to each request with a response message, the order is not guaranteed.
Sending requests uses a [Promise](https://github.com/reactphp/promise)-based interface that makes it easy to react to when a request is fulfilled (i.e. either successfully resolved or rejected with an error).

```php
$client->fetchFile($path)->then(
    function ($contents) {
        // file contents received
    },
    function (Exception $e) {
        // an error occured while executing the request
    }
});
```

If this looks strange to you, you can also use the more traditional [blocking API](#blocking).

#### Blocking

As stated above, this library provides you a powerful, async API by default.

If, however, you want to integrate this into your traditional, blocking environment,
you should look into also using [clue/block-react](https://github.com/clue/php-block-react).

The resulting blocking code could look something like this:

```php
use Clue\React\Block;

$loop = React\EventLoop\Factory::create();
$browser = new Clue\React\Buzz\Browser($loop);

$client = new Client($browser->withBase($uri /* change me */));
$promise = $client->fetchFile($path /* change me */);

try {
    $contents = Block\await($promise, $loop);
    // file contents received
} catch (Exception $e) {
    // an error occured while executing the request
}
```

Refer to [clue/block-react](https://github.com/clue/php-block-react#readme) for more details.

#### Streaming

The following API endpoint resolves with the file contents as a string:

```php
$client->fetchFile($path);
````

Keep in mind that this means the whole string has to be kept in memory.
This is easy to get started and works reasonably well for smaller files.

For bigger files it's usually a better idea to use a streaming approach,
where only small chunks have to be kept in memory.
This works for (any number of) files of arbitrary sizes.

The following API endpoint complements the default Promise-based API and returns
an instance implementing `ReadableStreamInterface` instead:

```php
$stream = $client->fetchFileStream($path);

$stream->on('data', function ($chunk) {
    echo $chunk;
});

$stream->on('error', function (Exception $error) {
    echo 'Error: ' . $error->getMessage() . PHP_EOL;
});

$stream->on('close', function () {
    echo '[DONE]' . PHP_EOL;
});
```

## Install

The recommended way to install this library is [through Composer](http://getcomposer.org).
[New to Composer?](http://getcomposer.org/doc/00-intro.md)

This will install the latest supported version:

```bash
$ composer require clue/viewvc-api-react:^0.4
```

See also the [CHANGELOG](CHANGELOG.md) for details about version upgrades.

## License

MIT


CHANGELOG.md
# Changelog

## 0.4.0 (2016-04-16)

* Feature: Add streaming API for fetching larger files
  (#20 by @clue)

  ```php
  $stream = $client->fetchFileStream($path);
  
  $stream->on('data', function ($chunk) {
      echo $chunk;
  });
  
  $stream->on('error', function (Exception $error) {
      echo 'Error: ' . $error->getMessage() . PHP_EOL;
  });
  
  $stream->on('close', function () {
      echo '[DONE]' . PHP_EOL;
  });
  ```

* Update dependencies, more SOLID code base

## 0.3.0 (2015-09-04)

* Feature / BC break: Build properly escaped URIs via Browser
  ([#17](https://github.com/clue/php-viewvc-api-react/pull/17))

  ```php
  // old
  $client = new Client($uri, $browser);
  
  // new
  $client = new Client($browser->withBase($uri));
  ```

* Update dependencies, more SOLID code base

## 0.2.0 (2015-04-10)

* Feature: Add `Client::fetchLog()` method
  ([#14](https://github.com/clue/php-viewvc-api-react/pull/14))

* Feature: Support including deleted files (attic) in CVS directory listings
  ([#11](https://github.com/clue/php-viewvc-api-react/pull/11))

* Feature: Support mixed UTF-8 and ISO-8859-1 (Latin1) encodings
  ([#12](https://github.com/clue/php-viewvc-api-react/pull/12))

* Fix: Improve parsing broken XHTML structure
  ([#10](https://github.com/clue/php-viewvc-api-react/pull/10), [#13](https://github.com/clue/php-viewvc-api-react/pull/13))

## 0.1.1 (2015-04-07)

* Fix: Accessing deleted files in older revisions now works
  ([#6](https://github.com/clue/php-viewvc-api-react/pull/6))

* Fix: Improve parsing XHTML structure
  ([#7](https://github.com/clue/php-viewvc-api-react/pull/7))

* Fix: Add trailing slash to base URL
  ([#5](https://github.com/clue/php-viewvc-api-react/pull/5))

* Improve test suite and directory listing example
  ([#8](https://github.com/clue/php-viewvc-api-react/pull/8), [#9](https://github.com/clue/php-viewvc-api-react/pull/9))

## 0.1.0 (2015-04-06)

* First tagged release


tests/bootstrap.php
<?php

require __DIR__ . '/../vendor/autoload.php';

class TestCase extends PHPUnit_Framework_TestCase
{
    protected function expectCallableOnce()
    {
        $mock = $this->createCallableMock();

        $mock
            ->expects($this->once())
            ->method('__invoke');

        return $mock;
    }

    protected function expectCallableOnceWith($value)
    {
        $mock = $this->createCallableMock();

        $mock
            ->expects($this->once())
            ->method('__invoke')
            ->with($this->equalTo($value));

        return $mock;
    }

    protected function expectCallableNever()
    {
        $mock = $this->createCallableMock();
        $mock
            ->expects($this->never())
            ->method('__invoke');

        return $mock;
    }

    /**
     * @link https://github.com/reactphp/react/blob/master/tests/React/Tests/Socket/TestCase.php (taken from reactphp/react)
     */
    protected function createCallableMock()
    {
        return $this->getMock('CallableStub');
    }

    protected function expectPromiseResolve($promise)
    {
        $this->assertInstanceOf('React\Promise\PromiseInterface', $promise);

        $promise->then($this->expectCallableOnce(), $this->expectCallableNever());

        return $promise;
    }

    protected function expectPromiseResolveWith($value, $promise)
    {
        $this->assertInstanceOf('React\Promise\PromiseInterface', $promise);

        $promise->then($this->expectCallableOnceWith($value), $this->expectCallableNever());

        return $promise;
    }

    protected function expectPromiseReject($promise)
    {
        $this->assertInstanceOf('React\Promise\PromiseInterface', $promise);

        $promise->then($this->expectCallableNever(), $this->expectCallableOnce());

        return $promise;
    }
}

class CallableStub
{
    public function __invoke()
    {
    }
}


tests/ClientTest.php
<?php

use Clue\React\ViewVcApi\Client;
use React\Promise;
use Clue\React\Buzz\Browser;
use Psr\Http\Message\RequestInterface;
use RingCentral\Psr7\Response;

class ClientTest extends TestCase
{
    private $uri = 'http://viewvc.example.org/';
    private $sender;
    private $client;

    public function setUp()
    {
        $this->sender = $this->getMockBuilder('Clue\React\Buzz\Io\Sender')->disableOriginalConstructor()->getMock();

        $browser = new Browser($this->getMock('React\EventLoop\LoopInterface'), $this->sender);

        $this->client = new Client($browser->withBase($this->uri));
    }

    public function testInvalidDirectory()
    {
        $promise = $this->client->fetchDirectory('invalid');
        $this->expectPromiseReject($promise);
    }

    public function testInvalidFile()
    {
        $promise = $this->client->fetchFile('invalid/');
        $this->expectPromiseReject($promise);
    }

    public function testFetchFile()
    {
        $response = new Response(200, array(), '# hello', '1.0', 'OK');

        $this->expectRequest($this->uri . 'README.md?view=co')->will($this->returnValue(Promise\resolve($response)));

        $promise = $this->client->fetchFile('README.md');

        $this->expectPromiseResolveWith('# hello', $promise);
    }

    public function testFetchFileStream()
    {
        $response = new Response(200, array(), '# hello', '1.0', 'OK');

        $this->expectRequest($this->uri . 'README.md?view=co')->will($this->returnValue(Promise\reject()));

        $stream = $this->client->fetchFileStream('README.md');

        $this->assertInstanceOf('React\Stream\ReadableStreamInterface', $stream);
    }

    /**
     * @expectedException InvalidArgumentException
     */
    public function testInvalidFileStream()
    {
        $this->client->fetchFileStream('invalid/');
    }

    public function testFetchFileExcessiveSlashesAreIgnored()
    {
        $this->expectRequest($this->uri . 'README.md?view=co')->will($this->returnValue(Promise\reject()));

        $promise = $this->client->fetchFile('/README.md');

        $this->expectPromiseReject($promise);
    }

    public function testFetchFileRevision()
    {
        $this->expectRequest($this->uri . 'README.md?view=co&pathrev=1.0')->will($this->returnValue(Promise\reject()));

        $promise = $this->client->fetchFile('/README.md', '1.0');

        $this->expectPromiseReject($promise);
    }

    public function testFetchDirectoryRevision()
    {
        $this->expectRequest($this->uri . 'directory/?pathrev=1.0')->will($this->returnValue(Promise\reject()));

        $promise = $this->client->fetchDirectory('/directory/', '1.0');

        $this->expectPromiseReject($promise);
    }

    public function testFetchDirectoryAttic()
    {
        $this->expectRequest($this->uri . 'directory/?hideattic=0')->will($this->returnValue(Promise\reject()));

        $promise = $this->client->fetchDirectory('/directory/', null, true);

        $this->expectPromiseReject($promise);
    }

    public function testFetchDirectoryRevisionAttic()
    {
        $this->expectRequest($this->uri . 'directory/?pathrev=1.1&hideattic=0')->will($this->returnValue(Promise\reject()));

        $promise = $this->client->fetchDirectory('/directory/', '1.1', true);

        $this->expectPromiseReject($promise);
    }

    public function testFetchLogRevision()
    {
        $this->expectRequest($this->uri . 'README.md?view=log&pathrev=1.0')->will($this->returnValue(Promise\reject()));

        $promise = $this->client->fetchLog('/README.md', '1.0');

        $this->expectPromiseReject($promise);
    }

    public function testFetchPatch()
    {
        $this->expectRequest($this->uri . 'README.md?view=patch&r1=1.0&r2=1.1')->will($this->returnValue(Promise\reject()));

        $promise = $this->client->fetchPatch('/README.md', '1.0', '1.1');

        $this->expectPromiseReject($promise);
    }

    private function expectRequest($uri)
    {
        $that = $this;

        return $this->sender->expects($this->once())->method('send')->with($this->callback(function (RequestInterface $request) use ($that, $uri) {
            $that->assertEquals('GET', $request->getMethod());
            $that->assertEquals($uri, $request->getUri());

            return true;
        }));
    }
}


tests/Io/LoaderTest.php
<?php

use Clue\React\ViewVcApi\Io\Loader;

class LoaderTest extends TestCase
{
    private $loader;

    public function setUp()
    {
        $this->loader = new Loader();
    }

    /**
     * @dataProvider xmlFiles
     * @param string $path
     */
    public function testValidXmlFixtures($path)
    {
        $xml = $this->loader->loadXmlFile(__DIR__ . '/../fixtures/' . $path);
    }

    public function xmlFiles()
    {
        return array_filter(array_map(
            function ($path) {
                return (substr($path, -5) === '.html') ? array($path) : null;
            },
            scandir(__DIR__ . '/../fixtures/')
        ));
    }

    public function testHtmlEntities()
    {
        $str = '<p>&auml;&hellip;&nbsp;&copy;</p>';
        $xml = $this->loader->loadXmlString($str);

        // c3 a4 e2 80 a6 c2 a0 c2 a9
        $this->assertEquals('ä… ©', (string)$xml);
    }

    public function testMixedEncodings()
    {
        // mixed UTF-8 and ISO-8859-1
        $str = "<p>ä and \xFC</p>";
        $xml = $this->loader->loadXmlString($str);

        $this->assertEquals('ä and ü', (string)$xml);
    }

    public function testLoadInvalidMarkupInputNotClosed()
    {
        $str = '<input type="hidden">';
        $xml = $this->loader->loadXmlString($str);

        $this->assertEquals('hidden', (string)$xml['type']);
    }

    public function testPrepareInvalidMarkupBrNotClosed()
    {
        $html = '<br>';
        $xml = $this->loader->loadXmlString($html);
    }

    public function testLoadInvalidMarkupSelectedAttributeNoValue()
    {
        $str = '<option selected>this</option>';
        $xml = $this->loader->loadXmlString($str);

        $this->assertEquals('selected', (string)$xml['selected']);
    }

    public function testLoadRemovedNavHeader()
    {
        $str = '<body><div class="vc_navheader"> invalid <asd> </okay> </div> </body>';
        $xml = $this->loader->loadXmlString($str);

        $this->assertFalse(isset($xml->div));
        $this->assertEquals(0, count($xml));
    }
}


src/Client.php
<?php

namespace Clue\React\ViewVcApi;

use RuntimeException;
use UnderflowException;
use SimpleXMLElement;
use InvalidArgumentException;
use Clue\React\Buzz\Browser;
use Psr\Http\Message\ResponseInterface;
use React\Promise;
use Clue\React\ViewVcApi\Io\Parser;
use Clue\React\ViewVcApi\Io\Loader;
use Rize\UriTemplate;
use Clue\React\Promise\Stream;

class Client
{
    private $brower;
    private $parser;
    private $loader;
    private $uri;

    public function __construct(Browser $browser, Parser $parser = null, Loader $loader = null, UriTemplate $uri = null)
    {
        if ($parser === null) {
            $parser = new Parser();
        }
        if ($loader === null) {
            $loader = new Loader();
        }
        if ($uri === null) {
            $uri = new UriTemplate();
        }

        // TODO: do not follow redirects
//         $browser = $this->browser->withOptions(array(
//             'follow_redirects' => false
//         ));

        $this->browser = $browser;
        $this->parser = $parser;
        $this->loader = $loader;
        $this->uri = $uri;
    }

    public function fetchFile($path, $revision = null)
    {
        if (substr($path, -1) === '/') {
            return Promise\reject(new InvalidArgumentException('File path MUST NOT end with trailing slash'));
        }

        // TODO: fetching a directory redirects to path with trailing slash
        // TODO: status returns 200 OK, but displays an error message anyways..
        // TODO: see not-a-file.html
        // TODO: reject all paths with trailing slashes

        return $this->fetch(
            $this->uri->expand(
                '{+path}?view=co{&pathrev}',
                array(
                    'path' => $path,
                    'pathrev' => $revision
                )
            )
        );
    }

    /**
     * Reads the file contents of the given file path as a readable stream
     *
     * This works for files of arbitrary sizes as only small chunks have to
     * be kept in memory. The resulting stream is a well-behaving readable stream
     * that will emit the normal stream events.
     *
     * @param string      $path
     * @param string|null $revision
     * @return ReadableStreamInterface
     * @throws InvalidArgumentException
     * @see self::fetchFile()
     */
    public function fetchFileStream($path, $revision = null)
    {
        if (substr($path, -1) === '/') {
            throw new InvalidArgumentException('File path MUST NOT end with trailing slash');
        }

        // TODO: fetching a directory redirects to path with trailing slash
        // TODO: status returns 200 OK, but displays an error message anyways..
        // TODO: see not-a-file.html
        // TODO: reject all paths with trailing slashes

        return Stream\unwrapReadable(
            $this->browser->withOptions(array('streaming' => true))->get(
                $this->uri->expand(
                    '{+path}?view=co{&pathrev}',
                    array(
                        'path' => $path,
                        'pathrev' => $revision
                    )
                )
            )->then(function (ResponseInterface $response) {
                // the body implements ReadableStreamInterface, so let's just return this to the unwrapper
                return $response->getBody();
            })
        );
    }

    public function fetchDirectory($path, $revision = null, $showAttic = false)
    {
        if (substr($path, -1) !== '/') {
            return Promise\reject(new InvalidArgumentException('Directory path MUST end with trailing slash'));
        }

        // TODO: path MUST end with trailing slash
        // TODO: accessing files will redirect to file with relative location URL (not supported by clue/buzz-react)

        return $this->fetchXml(
            $this->uri->expand(
                '{+path}{?pathrev,hideattic}',
                array(
                    'path' => $path,
                    'pathrev' => $revision,
                    'hideattic' => $showAttic ? '0' : null
                )
            )
        )->then(function (SimpleXMLElement $xml) {
            // TODO: reject if this is a file, instead of directory => contains "Log of" instead of "Index of"
            // TODO: see is-a-file.html

            return $xml;
        })->then(array($this->parser, 'parseDirectoryListing'));
    }

    public function fetchPatch($path, $r1, $r2)
    {
        return $this->fetch(
            $this->uri->expand(
                '{+path}?view=patch{&r1,r2}',
                array(
                    'path' => $path,
                    'r1' => $r1,
                    'r2' => $r2
                )
            )
        );
    }

    public function fetchLog($path, $revision = null)
    {
        // TODO: invalid revision shows error page, but HTTP 200 OK

        return $this->fetchXml(
            $this->uri->expand(
                '{+path}?view=log{&pathrev}',
                array(
                    'path' => $path,
                    'pathrev' => $revision
                )
            )
        )->then(array($this->parser, 'parseLogEntries'));
    }

    public function fetchRevisionPrevious($path, $revision)
    {
        return $this->fetchAllPreviousRevisions($path)->then(function ($revisions) use ($revision) {
            if (!isset($revisions[$revision])) {
                throw new UnderflowException('Unable to find previous version of given revision');
            }

            return $revisions[$revision];
        });
    }

    public function fetchAllPreviousRevisions($path)
    {
        return $this->fetchLogXml($path)->then(array($this->parser, 'parseLogRevisions'));
    }

    private function fetchLogXml($path)
    {
        return $this->fetchXml(
            $this->uri->expand(
                '{+path}?view=log',
                array(
                    'path' => $path
                )
            )
        );
    }

    private function fetchXml($url)
    {
        return $this->fetch($url)->then(array($this->loader, 'loadXmlString'));
    }

    private function fetch($url)
    {
        return $this->browser->get($url)->then(
            function (ResponseInterface $response) {
                return (string)$response->getBody();
            },
            function ($error) {
                throw new RuntimeException('Unable to fetch from ViewVC', 0, $error);
            }
        );
    }
}


src/Io/Loader.php
<?php

namespace Clue\React\ViewVcApi\Io;

use SimpleXMLElement;

class Loader
{
    private $entities;

    public function __construct(array $entities = null)
    {
        if ($entities === null) {
            // get all HTML entities (minus those for XML parsing)
            $entities = get_html_translation_table(HTML_ENTITIES, ENT_NOQUOTES, 'UTF-8');
            unset($entities['<'], $entities['>'], $entities['&']);
        }

        $this->entities = $entities;
    }

    public function loadXmlFile($path)
    {
        return $this->loadXmlString(file_get_contents($path));
    }

    public function loadXmlString($html)
    {
        // log output often uses garbled ISO-8859-1 and UTF-8 encodings
        $html = \ForceUTF8\Encoding::toUTF8($html);

        // fix invalid markup of outdated ViewVC versions
        // - help link in footer not terminated
        // - selected branch/tag in CVS "sticky tag" dropdown has not attribute value
        // - self closing elements with no trailing slash
        // - remove navheader because of its often incomplete form tags
        $html = str_replace('Help</strong></td>', 'Help</a></strong></td>', $html);
        $html = str_replace('selected>', 'selected="selected">', $html);
        $html = preg_replace('#<((?:input|br|hr|img)[^\/\>]*)>#', '<$1 />', $html);
        $html = preg_replace('#\<div class\=\"vc_navheader\"\>.*?\<\/div\>#s','', $html);

        // replace named HTML entities with their UTF-8 value
        $html = str_replace(array_values($this->entities), array_keys($this->entities), $html);

        // clean up namespace declaration
        $html = str_replace('xmlns="', 'ns="', $html);

        return new SimpleXMLElement($html);
    }
}


examples/directory.php
<?php

use Clue\React\ViewVcApi\Client;
use React\EventLoop\Factory as LoopFactory;
use Clue\React\Buzz\Browser;
use React\Stream\Stream;

require __DIR__ . '/../vendor/autoload.php';

$url = 'https://svn.apache.org/viewvc/';
$path = isset($argv[1]) ? $argv[1] : '/';
$revision = isset($argv[2]) ? $argv[2] : null;

$loop = LoopFactory::create();

// $dns = '10.52.166.2';
// $resolver = new React\Dns\Resolver\Factory();
// $resolver = $resolver->createCached($dns, $loop);
// $connector = new React\SocketClient\Connector($loop, $resolver);

// $socks = new \Clue\React\Socks\Client($loop, '127.0.0.1', 9050);
// $socks->setResolveLocal(false);
// $connector = $socks->createConnector();

// $sender = Clue\React\Buzz\Io\Sender::createFromLoopConnectors($loop, $connector);
$sender = null;

$browser = new Browser($loop, $sender);

$client = new Client($browser->withBase($url));

if (substr($path, -1) === '/') {
    $client->fetchDirectory($path, $revision)->then('print_r', 'printf');
} else {
    //$client->fetchFile($path, $revision)->then('print_r', 'printf');

    $stream = $client->fetchFileStream($path, $revision);

    // any errors
    $stream->on('error', 'printf');

    // pipe stream into STDOUT
    $out = new Stream(STDOUT, $loop);
    $out->pause();
    $stream->pipe($out);
}

$loop->run();


src/Io/Parser.php
<?php

namespace Clue\React\ViewVcApi\Io;

use SimpleXMLElement;

class Parser
{
    public function parseDirectoryListing(SimpleXMLElement $xml)
    {
        $files = array();

        // iterate over all anchor elements with an href attribute
        foreach ($xml->xpath('//tr/td/a[@name]') as $anchor) {
            $name = (string)$anchor['name'];

            // append slash to directory names
            $href = (string)$anchor['href'];
            if (substr($href, -1) === '/' || strpos($href, '/?') !== false) {
                $name .= '/';
            }

            $files []= $name;
        }

        return $files;
    }

    public function parseLogRevisions(SimpleXMLElement $xml)
    {
        $revisions = array();

        // iterate over all anchor elements with an href attribute
        foreach ($xml->xpath('//a[@href]') as $anchor) {
            // text label of anchor element
            $text = trim((string)$anchor);

            // only look for links to previous revision
            if (substr($text, 0, 8) !== 'previous') {
                continue;
            }

            // href contains r1 and r2 as query parameters
            $args = $this->linkParameters((string)$anchor['href']);

            // all links containing r2 are links to previous revision
            if (isset($args['r2'])) {
                $revisions[$args['r2']] = $args['r1'];
            }
        }

        return $revisions;
    }

    /**
     * Parse log entries from given XML document
     *
     * @param SimpleXMLElement $xml
     * @throws \UnexpectedValueException
     * @return array
     * @link https://gforge.inria.fr/scm/viewvc/viewvc.org/template-authoring-guide.html#variables-log
     */
    public function parseLogEntries(SimpleXMLElement $xml)
    {
        $entries = array();

        foreach ($xml->xpath('//div[pre]') as $div) {
            /* @var $div SimpleXMLElement */

            // skip "(vendor branch)" "em" tag if found
            $off = ((string)$div->em[0] === '(vendor branch)') ? 1 : 0;

            $entry = array(
                // revision is first "strong" element (subversion wraps this in "a" element)
                'revision' => (string)$this->first($div->xpath('.//strong[1]')),
                // date is in first "em" element
                'date' => new \DateTime((string)$div->em[0 + $off]),
                // author is in second "em" element
                'author' => (string)$div->em[1 + $off],
                // message is in only "pre" element
                'message' => (string)$div->pre
            );

            // ease parsing each line by splitting on "br" element, skip static rows for revision/date
            $parts = explode('<br />', substr($div->asXML(), 5, -6));
            unset($parts[0], $parts[1]);

            foreach ($parts as $part) {
                $part = new SimpleXMLElement('<body>' . $part . '</body>');
                $str = (string)$part;

                if (substr($str, 0, 7) === 'Diff to') {
                    $value = array();

                    foreach ($part->xpath('.//a') as $a) {
                        $text = (string)$a;
                        $pos = strrpos($text, ' ');

                        // text should be "previous X.Y", otherwise ignore "(colored)" with no blank
                        if ($pos !== false) {
                            $value[substr($text, 0, $pos)] = substr($text, $pos + 1);
                        }
                    }

                    $entry['diff'] = $value;
                } elseif (substr($str, 0, 7) === 'Branch:' || substr($str, 0, 9) === 'CVS Tags:' || substr($str, 0, 17) === 'Branch point for:') {
                    $value = array();

                    foreach ($part->xpath('.//a/strong') as $a) {
                        $value []= (string)$a;
                    }

                    $key = $str[0] === 'B' ? ($str[6] === ':' ? 'branches' : 'branchpoints') : 'tags';
                    $entry[$key] = $value;
                } elseif (substr($str, 0, 13) === 'Changes since') {
                    // "strong" element contains "X.Y: +1 -2 lines"
                    $value = (string)$part->strong;
                    $pos = strpos($value, ':');

                    // previous revision is before colon
                    $entry['previous'] = substr($value, 0, $pos);

                    // changes are behind colon
                    $entry['changes'] = substr($value, $pos + 2);
                } elseif (substr($str, 0, 14) === 'Original Path:') {
                    $entry['original'] = (string)$part->a->em;
                } elseif (substr($str, 0, 12) === 'File length:') {
                    $entry['size'] = (int)substr($str, 13);
                } elseif (isset($part->strong->em) && (string)$part->strong->em === 'FILE REMOVED') {
                    $entry['deleted'] = true;
                }
            }

            // previous is either set via "changes since" or link to "diff to" previous
            if (isset($entry['diff']['previous'])) {
                $entry['previous'] = $entry['diff']['previous'];
            }

            if ($off) {
                $entry['vendor'] = true;
            }

            $entries []= $entry;
        }

        return $entries;
    }

    private function first(array $a)
    {
        return $a[0];
    }

    private function linkParameters($href)
    {
        $args = array();
        $pos = strpos($href, '?');

        if ($pos !== false) {
            parse_str(substr($href, $pos + 1), $args);
        }

        return $args;
    }
}
